crates/amaru-mempool/src/strategies/in_memory_mempool.rs

```diff
@@ -0,0 +1,344 @@
+        }
+    }
+}
+
+impl<Tx> InMemoryMempool<Tx> {
+    pub fn new(config: MempoolConfig, tx_validator: Arc<dyn CanValidateTransactions<Tx>>) -> Self {
+        InMemoryMempool {
+            config,
+            inner: parking_lot::RwLock::new(MempoolInner::default()),
+            notify: Notify::new(),
+            tx_validator,
+        }
+    }
+
+    pub fn from_config(config: MempoolConfig) -> Self {
+        Self::new(config, Arc::new(DefaultCanValidateTransactions))
+    }
+}
+
+/// A default transactions validator.
+#[derive(Clone, Debug, Default)]
+pub struct DefaultCanValidateTransactions;
+
+impl<Tx> CanValidateTransactions<Tx> for DefaultCanValidateTransactions {
+    fn validate_transaction(&self, _tx: &Tx) -> Result<(), TransactionValidationError> {
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct MempoolInner<Tx> {
+    next_seq: u64,
+    entries_by_id: BTreeMap<TxId, MempoolEntry<Tx>>,
+    entries_by_seq: BTreeMap<MempoolSeqNo, TxId>,
+}
+
+impl<Tx> Default for MempoolInner<Tx> {
+    fn default() -> Self {
+        MempoolInner {
+            next_seq: 1,
+            entries_by_id: Default::default(),
+            entries_by_seq: Default::default(),
+        }
+    }
+}
+
+impl<Tx: Encode<()> + Clone> MempoolInner<Tx> {
+    /// Inserts a new transaction into the mempool.
+    /// The transaction id is a hash of the transaction body.
+    fn insert(
+        &mut self,
+        config: &MempoolConfig,
+        tx: Tx,
+        tx_origin: TxOrigin,
+    ) -> Result<(TxId, MempoolSeqNo), TxRejectReason> {
+        if let Some(max_txs) = config.max_txs
+            && self.entries_by_id.len() >= max_txs
+        {
+            return Err(TxRejectReason::MempoolFull);
+        }
+
+        let tx_id = TxId::from(&tx);
+        if self.entries_by_id.contains_key(&tx_id) {
+            return Err(TxRejectReason::Duplicate);
+        }
+
+        let seq_no = MempoolSeqNo(self.next_seq);
+        self.next_seq += 1;
+
+        let entry = MempoolEntry {
+            seq_no,
+            tx_id: tx_id.clone(),
+            tx,
+            origin: tx_origin,
+        };
+
+        self.entries_by_id.insert(tx_id.clone(), entry);
+        self.entries_by_seq.insert(seq_no, tx_id.clone());
+        Ok((tx_id, seq_no))
+    }
+
+    /// Retrieves a transaction by its id.
+    fn get_tx(&self, tx_id: &TxId) -> Option<Tx> {
+        self.entries_by_id.get(tx_id).map(|entry| entry.tx.clone())
+    }
+
+    /// Retrieves all the transaction ids since a given sequence number, up to a limit.
+    fn tx_ids_since(&self, from_seq: MempoolSeqNo, limit: u16) -> Vec<(TxId, u32, MempoolSeqNo)> {
+        let mut result: Vec<(TxId, u32, MempoolSeqNo)> = self
+            .entries_by_seq
+            .range(from_seq..)
+            .take(limit as usize)
+            .map(|(seq, tx_id)| {
+                if let Some(entry) = self.entries_by_id.get(tx_id) {
+                    let tx_size = to_cbor(&entry.tx).len() as u32;
```

Comment:
even if this uses the serialization buffer to avoid allocations, how about storing the size in the map entry as well?